python/amici/ode_export.py

```diff
@@ -2146,4 +2525,86 @@ def sanitize_basic_sympy(basic):
         # if this expression contains any pysb.Components, we can
         # safely apply str() to the full expression as str will do
         # the right thing here
-        return sp.sympify(str(basic))
+
+        # ensure that pysb symbols are correctly interpreted as symbols and
+        # not as functions etc
+        locals = {
+            fs.name: sp.Symbol(fs.name)
+            for fs in list(basic.expr_free_symbols)
+            if isinstance(fs, pysb.core.Component)
+        }
+        return sp.sympify(str(basic), locals=locals)
+
+
+def get_function_defition(fun, name):
+    """Constructs the function definition for a given function
+
+    Arguments:
+        fun: function name @type str
+        name: model name @type str
+
+    Returns:
+    c++ function definition string
+
+    Raises:
+
+    """
+    return \
+        f'extern void {fun}_{name}{functions[fun]["signature"]};'
+
+
+def get_function_implementation(fun, name):
+    """Constructs the function implementation for a given function
+
+    Arguments:
+        fun: function name @type str
+        name: model name @type str
+
+    Returns:
+    c++ function implementation string
+
+    Raises:
+
+    """
+    return \
+        '{ind4}virtual void f{fun}{signature} override {{\n' \
+        '{ind8}{fun}_{name}{eval_signature};\n' \
+        '{ind4}}}\n'.format(
+            ind4=' '*4,
+            ind8=' '*8,
+            fun=fun,
+            name=name,
+            signature=functions[fun]["signature"],
+            eval_signature=remove_typedefs(functions[fun]["signature"])
+        )
+
+
+def remove_typedefs(signature):
+    """Strips typedef info from a function signature
+
+    Arguments:
+        signature: function signature @type str
+
+    Returns:
+    string that can be used to construct function calls with the same
+    variable names and ordering as in the function signature
+
+    Raises:
+
+    """
+    typedefs = [
```

Comment:
Inconsistent final space in contained strings. Intended?

Reply:
In parts, will add explanatory comments (the `*` is for pointers). Whitespace after `int` was indeed missing.